### ISSUE TITLE
gradient of eigh does not respect `symmetrize_input`

### ISSUE THREAD

**Issue by jakevdp** [google/jax#12795]
### Description

The JVP rule for `lax.linalg.eigh` effectively hard-codes `symmetrize_inputs=True`, which leads to incorrect gradients when `symmetrize_inputs=False`. Here is a quick repro: first we construct a lower-triangular matrix and its explicitly symmetric counterpart, and confirm that they have the same eigenvalues:
```python
import jax.numpy as jnp
import jax
import numpy as np
rng = np.random.RandomState(0)
N = 3

# Construct a lower-triangular matrix
x_L = np.tril(rng.randn(N, N) + 1j * rng.randn(N, N))

# Construct the Hermitian matrix represented by x_L
x_sym = x_L + x_L.conj().T
x_sym[np.diag_indices(N)] /= 2  # correct for doubling of diagonal

def eigvalsh_lower(x):
  return jax.lax.linalg.eigh(x, lower=True, symmetrize_input=False)[1]

print(eigvalsh_lower(x_L))
# [-1.7833483  0.7766917  4.5350485]
print(eigvalsh_lower(x_sym))
# [-1.7833483  0.7766917  4.5350485]
```
Now we take the gradient with respect to the first eigenvalue. This should be effectively the same computation, and so the gradients should match and be lower-triangular, but they are not:
```python
def first_eigvalh_lower(x):
  return eigvalsh_lower(x)[0]

print(jax.grad(first_eigvalh_lower)(x_L))
# [[ 0.10143662+0.j         -0.07633214-0.09540727j -0.26702693+0.07010485j]
#  [-0.07633214+0.09540727j  0.14717706+0.j          0.13500279-0.30390963j]
#  [-0.26702693-0.07010485j  0.13500279+0.30390963j  0.75138617+0.j        ]]
print(jax.grad(first_eigvalh_lower)(x_sym))
# [[ 0.23979793+0.j         -0.21083055-0.16992857j -0.30570662+0.12455184j]
#  [-0.21083055+0.16992857j  0.30577928+0.j          0.1805161 -0.3261397j ]
#  [-0.30570662-0.12455184j  0.1805161 +0.3261397j   0.45442304+0.j        ]]
```
The issue comes from the fact that `lax.eig_p` does not have a `symmetrize_input` parameter, and the JVP rule symmetrizes the input unconditionally: https://github.com/google/jax/blob/188cf17d04bccd06801ada8202f4fa8e842a695d/jax/_src/lax/linalg.py#L714-L715

### What jax/jaxlib version are you using?

main branch

### Which accelerator(s) are you using?

CPU

### Additional system info

Python 3.9

**Comment by jakevdp** [comment]
It occurs to me that we should fix this before merging anything like https://github.com/google/jax/pull/12696, because that adds the `symmetrize_input` argument to `eigvalsh` and `cholesky`. Currently there's no way to get incorrect gradients for those because there's no way to set `symmetrize_input=False`.
